pkg/controllers/resourcesyncer/resourcesyncer.go

```diff
@@ -0,0 +1,153 @@
+package resourcesyncer
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/utils/strings/slices"
+
+	corecontrollers "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	"github.com/rancher/rancher/pkg/types/config"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// resourceSyncer synchronizes secrets between upstream and downstream clusters.
+type resourceSyncer struct {
+	upstreamSecretClient   corecontrollers.SecretInterface
+	downstreamSecretClient corecontrollers.SecretInterface
+
+	mgmtClusterName string
+	rkeControlPlane *types.NamespacedName
+}
+
+const (
+	SyncToCluster   = "provisioning.cattle.io/sync-to-cluster"
+	SyncToNamespace = "provisioning.cattle.io/sync-to-namespace"
+	SyncedTimestamp = "provisioning.cattle.io/last-synced-at"
+)
+
+// list of namespaces that are allowed to have resources synchronized between
+// the upstream and downstream cluster
+var allowedNamespaces = []string{"kube-system"}
+
+func Register(ctx context.Context, cluster *config.UserContext, mgmt *config.ScaledContext) {
+	// we only want to run this controller for non-local clusters
+	if cluster.ClusterName == "local" {
+		return
+	}
+
+	r := resourceSyncer{
+		upstreamSecretClient:   mgmt.Core.Secrets(""),
+		downstreamSecretClient: cluster.Core.Secrets(""),
+		mgmtClusterName:        cluster.ClusterName,
+	}
+
+	// populate the rkeControlPlane field so that it's available in the handler
+	controlPlanes, err := mgmt.Wrangler.RKE.RKEControlPlane().Cache().List("", labels.Everything())
+	if err != nil {
+		logrus.Warnf("failed to list RKEControlPlanes from cache: %v", err)
+		return
+	}
+
+	for _, plane := range controlPlanes {
+		if plane.Spec.ManagementClusterName == r.mgmtClusterName {
+			r.rkeControlPlane = &types.NamespacedName{
+				Name:      plane.Name,
+				Namespace: plane.Namespace,
+			}
+			break
+		}
+	}
+
+	if r.rkeControlPlane == nil {
+		logrus.Warnf("failed to find RKEControlPlane for cluster %v", r.mgmtClusterName)
+		return
+	}
+
+	for _, ns := range allowedNamespaces {
+		// TODO: all resources we want synced will need to register this handler (probably just secrets/configmaps for now)
+		mgmt.Core.Secrets(ns).Controller().AddHandler(ctx, "secret-sync", r.syncSecret)
```

Comment:
the handler should be labeled more appropriately like `resourcesyncer-secret-sync`

Reply:
theoretically we could use a tweaklist with a custom informer here to filter out the secrets that actually get enqueued, thus greatly reducing cycles. I'll have to dig up some more documentation around the tweaklist to see if that is something we actually want to do.